Update Twitter SDK from `Net::Twitter` to `Twitter::API`

bombsimon opened the pull request:
This will potentially fix https://github.com/metacpan/metacpan-web/issues/2322

I don't know how to best test this but I ran similar code in a separate script to confirm the SDK and API responses. I tried to setup [metacpan-docker](https://github.com/metacpan/metacpan-docker) but it crashes instantly under WSL. Any help testing this would be appreciated!

Thanks!

oalders commented on lib/MetaCPAN/Server/Controller/Login/Twitter.pm:
```diff
@@ -24,39 +24,46 @@ sub index : Path {
-                access_token_secret => $access_token_secret
+                id                  => $response->{user_id},
+                name                => $response->{screen_name},
+                access_token        => $response->{oauth_token},
```
I think, for security, we no longer need to store `oauth_token` and `oauth_token_secret`. I don't believe we actually use the tokens anywhere else in the API.  @mickeyn?

mickeyn replied:
I don't see that we do anywhere in the code.

oalders commented:
> I tried to setup [metacpan-docker](https://github.com/metacpan/metacpan-docker) but it crashes instantly under WSL

Was there a specific error message?

bombsimon commented:
> > I tried to setup [metacpan-docker](https://github.com/metacpan/metacpan-docker) but it crashes instantly under WSL
> 
> Was there a specific error message?

Sorry, I should've provided more info, I was mainly curious if that was your preferred way to test things like this.

The issue seems to be Windows->WSL specific with mounting volumes. I did a quick workaround by removing `read_only: true` from the `web` volumes which made the web frontend start. Let me do some more investigation on how to use my local `metacpan-api` to see if I can figure it out and I'll let you know if I need any assitance.

oalders commented:
Docker is a good way to test locally. Ideally we'd want a passing build on CircleCI as well. It's not quite working on forked PRs: #1072

We should fix that `read_only` issue. I think it only happens in some versions of `docker-compose`.

I'm not actually sure how good our test coverage is for Twitter OAuth. If you're able to confirm that there's coverage on your change, that would be helpful.

bombsimon commented:
I've tried both WSL and vanilla Windows with both Rancher Desktop and Docker Desktop now with the same result. First failing on `read_only` and when removing that volumes not mounting properly. F.ex. the compose file contains this for `api`:

```yaml
- type: bind
  source: ./src/metacpan-api
  target: /metacpan-api
```

But yet wen running it I get

```sh
$ docker-compose up api
metacpan_elasticsearch_1 is up-to-date
metacpan_pgdb_1 is up-to-date
metacpan_traefik_1 is up-to-date
Starting metacpan_api_1 ... done
Attaching to metacpan_api_1
api_1                     | exec /metacpan-api/wait-for-es.sh: no such file or directory
ERROR: Windows named pipe error: The pipe has been ended. (code: 109)
```

Since I normally never use Windows or WSL I have no clue what this is and I'm not sure it's worth trobuleshooting so I'll give this a go in a Linux environment instead.

Regarding (unit) testing there seem to be none for this but I'm not really sure what to test here. OAuth 2.0 is by nature hard to test since it contains the consent screen. I could write tests that assert my redirect URL is at the authorize endpoint for twitter and that it fails with missing token but I don't see a reasonable way to generate a working verifier in a unit test. Do you have any suggestions?

oalders commented:
@semifor do you have any suggestions on how to test this?

oalders commented:
> $ docker-compose up api

Just a small thing, but `docker-compose up api_test` will bring up an API that is better suited to running the test suite, assuming it actually starts up. 😄 

It looks like the `metacpan-api` docker container hasn't been updated in quite some time. I'll get a new image pushed to Docker hub in the next day or so. That might help. I'll update here when it's available.

bombsimon commented:
> > $ docker-compose up api
> 
> Just a small thing, but `docker-compose up api_test` will bring up an API that is better suited to running the test suite, assuming it actually starts up. 😄
> 
> It looks like the `metacpan-api` docker container hasn't been updated in quite some time. I'll get a new image pushed to Docker hub in the next day or so. That might help. I'll update here when it's available.

Oh, I didn't spot any difference. In fact, they seem to use the same image (`metacpan/metacpan-api:latest`) and the only difference being the `env_file`.

After removing `read_only` I got most of it working. I still can't create the API containers (any of them) getting this error:

```sh
api_test_1                | + exec -- /metacpan-api/wait-for-it.sh -t 15 -s pgdb:5432 -- 'morbo -l http://*:5000 -w app.psgi -w bin -w lib -w templates --verbose' ./bin/api.pl
api_test_1                | wait-for-it.sh: waiting 15 seconds for pgdb:5432
api_test_1                | wait-for-it.sh: pgdb:5432 is available after 0 seconds
api_test_1                | /metacpan-api/wait-for-it.sh: line 199: /metacpan-api/morbo -l http://*:5000 -w app.psgi -w bin -w lib -w templates --verbose: No such file or directory
```

Manually setting `command` to `morbo -l http://*:5000 -w app.psgi -w bin -w lib -w templates --verbose ./bin/api.pl` will make the API start

```sh
api_test_1                | Server available at http://127.0.0.1:5000
```

It's a bit complex to run this because I can't do it locally but i created a VM in GCP which I'm using and the redirect URI for the Twitter API is set to `http://localhost:5000` where the `api` is running (not `api_test`). To handle this I created an SSH-tunnel on port 5000 and expose `api_test` on 5000 instead of `api` but now when I try to log in I get my full GitHub profile but empty `access_token` and not sure if related but also `"looks_human" : false`.

I tried logging in with Twitter since it uses the same login flow as when linking but that fails because the Twitter credentials are invalid. `403 Forbidden at /usr/local/lib/perl5/site_perl/5.30.1/Twitter/API.pm line 367`.

I ended up configuring a personal client with a proper callback URL and I can now get into the flow and test it by logging in (although not linking yet). And I guess this was a good think because it doesn't seem to work as expected. Mainly the way cookies are set seems to be flaky, the cookie for the token doesn't refresh and the token secret doesn't get set. I need to do some troubleshooting to figure out how this works, it was a long time ago since I looked at `Catalyst` and `CGI::Simple::Cookie` so I'm a bit rusty 😸

oalders commented:
> Mainly the way cookies are set seems to be flaky,

I'm pretty sure that's the root cause of some of the general login problems with the site. Any insight on how to fix that would be great.

You may actually want to `docker build` the API image if you're in this deep already. That might fix some issues. The main difference between `api` and `api_test` is that `api_test` should be networked to an elasticsearch container with test data. That allows you to run tests without polluting your deployed containers.

Kudos to you for getting this going in GCP!

bombsimon commented:
> > Mainly the way cookies are set seems to be flaky,
> 
> I'm pretty sure that's the root cause of some of the general login problems with the site. Any insight on how to fix that would be great.

Yeah I think you're right! I saw you already use and configure `Catalyst::Plugin::Session` so I did a naive test and just replaced the usage of `cookie` with `session` and it seemed to work consistently for a while but I tried from a separate computer just to be sure which didn't work so now I'm back to where I started.

> You may actually want to `docker build` the API image if you're in this deep already. That might fix some issues. The main difference between `api` and `api_test` is that `api_test` should be networked to an elasticsearch container with test data. That allows you to run tests without polluting your deployed containers.

Alright, might be worth it but I'll start by trying to see if I can make the cookies more persistent or look for a different way to persist the credentials. I'll have limited time this week but hopefully I can spend some more time this weekend if no one beats me to it.

semifor commented:
I'm just catching up here. Is the login issue a Twitter login issue? There
are indeed some bugs in Twitter's OAuth authentication/authorization flow
and they have something to do with cookies or browser state. They've fixed
some but others persist and they aren't likely to fix them.



On Wed, Oct 26, 2022 at 1:24 PM Simon Sawert ***@***.***>
wrote:

> Mainly the way cookies are set seems to be flaky,
>
> I'm pretty sure that's the root cause of some of the general login
> problems with the site. Any insight on how to fix that would be great.
>
> Yeah I think you're right! I saw you already use and configure
> Catalyst::Plugin::Session so I did a naive test and just replaced the
> usage of cookie with session and it seemed to work consistently for a
> while but I tried from a separate computer just to be sure which didn't
> work so now I'm back to where I started.
>
> You may actually want to docker build the API image if you're in this
> deep already. That might fix some issues. The main difference between api
> and api_test is that api_test should be networked to an elasticsearch
> container with test data. That allows you to run tests without polluting
> your deployed containers.
>
> Alright, might be worth it but I'll start by trying to see if I can make
> the cookies more persistent or look for a different way to persist the
> credentials. I'll have limited time this week but hopefully I can spend
> some more time this weekend if no one beats me to it.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/metacpan/metacpan-api/pull/1071#issuecomment-1292613285>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAJFDZFQAYUUJ3KXSSGANTWFGHRLANCNFSM6AAAAAARNL2TJI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

bombsimon commented:
> Is the login issue a Twitter login issue? 

The root cause for me is that I can't link or sign in to Twitter via MetaCPAN.

I thought this could be related to `Net::Twitter` so I replaced with `Twitter::API`. It does not work in this PR due to how cookies are set to preserve token and token secret. They are set here: https://github.com/metacpan/metacpan-api/blob/1ef7cb02c8705e4a3699d585403a62975853284b/lib/MetaCPAN/Server/Controller/Login/Twitter.pm#L56-L59

But when read here they're outdated or missing: https://github.com/metacpan/metacpan-api/blob/1ef7cb02c8705e4a3699d585403a62975853284b/lib/MetaCPAN/Server/Controller/Login/Twitter.pm#L28-L30

This problem is the same in this PR and has nothing to do with with `Twitter::API` (or `Net::Twitter`) afaict.

For reference, this is the code I use to issue credentials, then just copy them from the screen and address bar and run the script again to obtain user tokens:

```perl
#!/usr/bin/env perl

use warnings;
use strict;
use feature qw( say );

use Data::Printer;
use Twitter::API;

sub main {
    my ( $oauth_token, $oauth_token_secret, $oauth_verifier ) = @ARGV;

    my $client = Twitter::API->new_with_traits(
        api_version     => '2',
        consumer_key    => '<redacted>',
        consumer_secret => '<redacted>',
    );

    if ( !$oauth_token ) {
        my $request  = $client->oauth_request_token( callback => 'http://127.0.0.1:5000' );
        my $auth_url = $client->oauth_authorization_url(
            {
                oauth_token => $request->{oauth_token},
            }
        );

        say $auth_url;
        say $request->{oauth_token};
        say $request->{oauth_token_secret};

        return 1;
    }

    my $data = $client->oauth_access_token(
        token        => $oauth_token,
        token_secret => $oauth_token_secret,
        verifier     => $oauth_verifier,
    );

    p $data;

    return 1;
}

main();
```

oalders commented:
There is now an updated `metacpan/metacpan-api` image on Docker Hub.

oalders commented:
> I'm just catching up here.

Initially the question was about finding a simple way to test Twitter in Catalyst, but I guess we've arrived at a larger problem to solve.

oalders commented:
Issue probably related to this: https://github.com/metacpan/metacpan-web/issues/2192

oalders commented:
If you rebase this PR I think we can get the CircleCI builds working.

semifor commented:
@bombsimon FWIW, you can simplify your process for obtaining OAuth tokens. You don't need to run a local server to receive the callback. If you don't pass a callback URL, Twitter::API sets it to 'oob', which results in Twitter displaying a PIN rather than redirecting to a callback URL.

Here's a modified version that uses the oob/PIN method:

```perl
#!/usr/bin/env perl

use warnings;
use strict;
use feature qw( say );

use Data::Printer;
use Twitter::API;

sub main {
    my $client = Twitter::API->new_with_traits(
        api_version     => '2',
        consumer_key    => $ENV{TWITTER_API_CONSUMER_KEY},
        consumer_secret => $ENV{TWITTER_API_CONSUMER_SECRET},
    );

    my $request  = $client->oauth_request_token;
    my $auth_url = $client->oauth_authorization_url(
        {
            oauth_token => $request->{oauth_token},
        }
    );

    my $oauth_token = $request->{oauth_token};
    my $oauth_token_secret = $request->{oauth_token_secret};

    print "
    Authorize with Twitter at: $auth_url
    Then enter the PIN number displayed in the browser: ";

    my $oauth_verifier = <STDIN>;
    chomp $oauth_verifier;

    my $data = $client->oauth_access_token(
        token        => $oauth_token,
        token_secret => $oauth_token_secret,
        verifier     => $oauth_verifier,
    );

    p $data;

    return 1;
}

main();
```

bombsimon commented:
@semifor thanks! Yeah I know, but since I don't think 'oob' and PIN will work in MetaCPAN as a final solution I wanted to try a flow as similar as possible. 

The state with Twitter right now is that everything is working except setting the cookies so I don't think ther's anything we need to change there. 

However regarding the original question, do you have any good way to unit test this? Do you do any unit testing with the actual API or have experience doing so?

semifor commented:
> I wanted to try a flow as similar as possible.

Ah! That makes sense.

I use Test::Spec for many of my tests and mock responses for API calls with Test::Spec::Mocks.

bombsimon commented:
I haven't spent any time on looking into a more stable local setup and thus I still cannot properly sign in with GitHub and link my account. However, when tinkering around and trying to do a MRE outside MetaCPAN I noticed that creating the session first seemed to make it more consistent. When I added this change to this PR I could sign in with Twitter in my local version running in Docker.

I don't know if I should spend time on a stable Docker setup that works for me so I can test the whole signing in and linking process or testing next. At least I don't think a test can prove that this will work, however if I figure out how to run a test with a persisted client and session with a mocked backend we can ensure that session data is stored properly. 

If anyone has a working local setup and want to give this a go that would be awesome! If not, let me know what you think a reasonable next step would be. Afaict the current Twitter implementation in production is broken for everyone, no?

Also I rebased on origin and CI is now green.

oalders commented on cpanfile:
```diff
@@ -161,6 +160,7 @@ requires 'Try::Tiny', '0.30';
 requires 'Type::Tiny', '2.000001';
 requires 'Types::Path::Tiny';
 requires 'Types::URI';
+requires 'Twitter::API', '1.0006';
```
After running `carton install` there should be an updated `cpanfile.snapshot` that you can commit.

bombsimon replied:
Not familiar with `carton` but I ran `carton install` and pushed the changes. Seems to be a lot not related to Twitter dependencies but maybe that's expected?

haarg commented:
Rebased and merged with d1b8c1980e8186571cfda380b8a918d116bd3072.

Thanks.
